### ISSUE TITLE
Redefined feature: "Creating actions" (conditionals)

### ISSUE THREAD

**Issue by christianalfoni** [christianalfoni/cerebral#46]
Okay, so we got some great input from @a-s-o. Inspired by [node-machines](http://node-machine.org/). The good thing here is that we keep the really simple syntax that we have now. We rather add a way to define custom paths and do type checking on the actions. I think this will be a life saver in larger projects and especially with 2+ teams ;-)

It is really important not to get fooled by the seemingly complexity here. Signals will work exactly the same, but you get added functionality for custom and typed actions.

### Default sync/async action
```js
const noPathAction = function (args, state, next) {
  next({foo: 'bar'}); // Always next, no return, to keep consistency
};

const defaultPathsAction = function (args, state, next) {
  if (conditional) {
    next.success({foo: 'bar'});
  } else {
    next.error({foo: 'bar'});
  }
};

// sync
controller.signal('appMounted', noPathAction);
controller.signal('appMounted', defaultPathsAction, {
  success: [],
  error: []
});

// async
controller.signal('appMounted', [noPathAction]);
controller.signal('appMounted', [defaultPathsAction, {
  success: [],
  error: []
}]);
```
So "success" and "error" are suggested default paths. Current "resolve" and "reject" should be changed. The reason is to avoid confusion with a promise. Yes, there is an underlying promise on async actions, but the `next` method chooses a path, it does not reject the actual promise. Any actual errors in the code (thrown errors) will be a rejected promise and will be thrown to console like normal and the signal will stop.

### Custom paths
```js
const customPathsAction = function (args, state, next) {
  if (conditional) {
    next.foo({foo: 'bar'});
    next({foo: 'bar}); // Default to foo
  } else {
    next.bar({foo: 'bar'});
  }
};
customPathsAction.defaultOutput = 'foo';
customPathsAction.outputs = ['foo', 'bar'];

// sync
controller.signal('appMounted', customPathsAction, {
  foo: [],
  bar: []
});

// async
controller.signal('appMounted', [customPathsAction, {
  foo: [],
  bar: []
}]);
```

### Type checked actions
```js
import {Types} from 'cerebral-some-package';

const noPathTypedAction = function (args, state, next) {
    next({foo: 'bar'})
};
noPathTypedAction.input = {
    foo: Types.String
};
noPathTypedAction.output = {
    foo: Types.String
};

const defaultPathsTypedAction = function (args, state, next) {
  if (conditional) {
    next.success({foo: 'bar'});
  } else {
    next.error({foo: 'bar'});
  }
};
defaultPathsTypedAction.input = {
    foo: Types.String,
    bar: Types.Bool
};
defaultPathsTypedAction.outputs = {
    success: {
      foo: Types.String
    },
    error: {
      foo: Types.String
    }
};

const customPathsTypedAction = function (args, state, next) {
  if (conditional) {
    next.foo({foo: 'bar'});
  } else {
    next.bar({foo: 'bar'});
  }
};
customPathsTypedAction.input = {
    foo: Types.String
};
customPathsTypedAction.outputs = {
    foo: {
      foo: Types.String
    },
    bar: {
      foo: Types.String
    }
};
```

Okay, so the thing here is that you commonly get away with default sync and async actions. You might need to reach out for a custom action to define some custom paths, but as you can see the definition is really simple.

In bigger projects and more people working on the project you can bring it up a notch, bringing some type checking into the mix. This will give errors when you refresh the page (compile time) as the signal can be analyzed and verified even before it runs. 

PS!
There is not need for top level array, we can use ES6 spread
```js
controller.signal('somethingHappened', syncAction, ...otherActions );
controller.signal('somethingHappened', [asyncAction], ...otherActions );
```

**Comment by a-s-o** [comment]
Good summary Christian. This would be very useful.  The only thing I would add is allow the ability to specify a `defaultOutput` override for actions that have multiple paths. 

Here is my test implementation for reference (https://gist.github.com/a-s-o/d9edfb76bea72bc5e1e2). Please let me know if you need something from me.

**Comment by idream3** [comment]
Looking really good :)

About the array at the top level we should be able to do this now using the spread operator.

```javascript
var otherActions = [
  action1,
  action2
]

controller.signal('somethingHappened', syncAction, ...otherActions );
controller.signal('somethingHappened', [asyncAction], ...otherActions );
```

**Comment by christianalfoni** [comment]
Ah, you already updated, great! I will add it to the implementation details. Great feedback @idream3

**Comment by christianalfoni** [comment]
@a-s-o Are you thinking that the default output would run with `next()` or that the action does not trigger a `next` at all? Thanks for reference. There is quite a bit of work to update documentation, demos etc. , but should not take too long :-)

**Comment by a-s-o** [comment]
Hi, to me the `defaultOutput` has two benefits.

1) It is declarative and easy to find out by looking at the action definition exactly what the action will do when `next` is called. Here is a short example:
```javascript
const upsertDocument = {
  inputs: { doc: Obj },
  outputs: {
    created: {},
    updated: {},
    error: {}
  },
  fn: function userLogin (args, state, next) {
    next.created(/* We know where this goes */);
    next.updated(/* We know which path this will go */);
    next(/* Where does next go?  */);
  }
};
```
We know from our knowledge that there is an additional output path `success` and that is where the last `next()` will output but having a `defaultOutput='updated'` is a little more explicit and flexible. Just personal preference.
2) By allowing the defaultOutput to be defined, we know where the user wants the return value for `sync` functions' to go, so we won't have to force they always call `next` from actions. They can continue using normal functions. `throw --goes to--> error` and `return --goes to--> success`

**Comment by a-s-o** [comment]
BTW, I can assist with the docs once you have an implementation down

**Comment by marbemac** [comment]
Love this. @christianalfoni it's similar to the PropTypes functionality I was describing in an earlier conversation, but better.

IMHO the biggest issue with signals/actions right now is that args is mutated arbitrarily during the life of a signal, and it can become hard to track which actions depend on what args, and which actions are changing the args. 

If we have an easy way to describe both the inputs and outputs of actions (and which inputs are required versus optional), Cerebral could, when a signal is added, loop through the actions and make sure that they fit together, and are compliant, based on the descriptions of input output for each signal. If Action 3 requires a "foo" argument, and neither Action 2 or 1 output a "foo" argument, and Action 1 does not require a "foo" argument, this chain is not compliant.

Not compliant:
```
Action 1 
  input: none required
  output: bar
Action 2
  input: bar required
  output: none
Action 3 
  input: foo required # foo is required and wasn't outputted in a prior action
  output: none
```

Compliant
```
# this is compliant because foo is required in the first action. We can assume the user will pass it in as the first action. Separate type checking can log an error if the user calls a signals and is missing required args.
Action 1 
  input: foo required
  output: bar
Action 2
  input: bar required
  output: none
Action 3 
  input: foo required
  output: none

# this is compliant because action 1 outputs foo
Action 1 
  input: none required
  output: bar, foo
Action 2
  input: bar required
  output: none
Action 3 
  input: foo required
  output: none
```

Basically, with this sort of system, we can enforce inter action dependencies, and make it very hard to put together actions in a way that breaks. As you mention, it also makes action dependencies more explicit, and thus easier to work with for teams of developers trying to reason about signals/actions for the first time.

**Comment by christianalfoni** [comment]
This is very interesting indeed! This "invisible" contract between the actions running can now be defined and controlled at compile time. Very very nice!

@a-s-o What do you think about adding inputs type checking too?

Btw, I understand how you think about defaultOutput, so consider that part of the spec :-)

**Comment by a-s-o** [comment]
Hi Christian,

I am all for input types as well. Looks like we already have them in your original comment, the draft implementation and our discussion in #45. I think they will work great together. If all actions specify inputs and outputs, it is clear what an action will need and what an action may output/overwrite in the args.

Furthermore since signals are just compositions of actions, it is then obvious what a signal's input will be and so they can all be checked together at call time. I never got that far in my composition but I think that can be implemented over time.

**Comment by a-s-o** [comment]
Along these lines, may I recommend tcomb (http://github.com/gcanti/tcomb). It is a really great library with a very small footprint (2kb gzipped) that can do a lot of the type checking and composition work for us.  

Specifically take a look at tcomb's `struct` functionality which can be used to specify action `inputs` and `outputs`. structs come with an `extend` method, so they can be easily composed. They throw when properties conflict. They also output frozen objects, so args cannot be directly overwritten in actions by accident. Properties will need to be specifically output/returned if overwriting is necessary. 

Finally, all the type checking's throwing/freezing gets automatically removed in production (`process.env.NODE_ENV === 'production'`) so it does not have any significant performance impact.

tcomb would complement cerebral very much and keep the type checking implementation external. +1 from me

**Comment by marbemac** [comment]
tcomb looks like a fantastic option, great recommendation @a-s-o!

**Comment by christianalfoni** [comment]
Hi guys!

I have started the implementation and should be done by tomorrow. I have hit one issue though, which I would like to discuss:

```js
controller.signal('appMounted',
  [async1, async2, async3],
  {
    success: [],
    error: []
  }
);
```
The three actions might resolve to different paths.... so what should happen? Should we allow resolving to multiple paths?

**Comment by a-s-o** [comment]
Hmm, I can see how that may get complex. Are you thinking of merging multiple paths in the final object from the various paths? 

Maybe only allow the error path in the final object and only allow multiple paths right after a particular action. Finally errors should stop the chain so, there should not be a case where multiple errors are accumulated, like `Promise.all()`.

```javascript
controller.signal('appMounted', ...[
  [
     async1,
     [async2, { some_path: [doSomethingElse], error: [recoverFromAsync2Error] }],
     [async3, { another_path: [doAnotherThing] }]
  ],
  setResults,
  { error: [ unhandledException ]  }
])
```
So basically, diverging paths are handled right beside where the action is added to the chain. If an action outputs on a particular path and it is not handled, then the result simple get merged into `args` and the chain continues on (again, sort of like `promise.then()`). So a pseudo-code example using promise api, is like this:
```javascript
controller.signal('appMounted', Promise.all([
   async1,
   async2.when('somePath', doSomethingElse).catch(recoverFromAsync2Error),
   async3.when('anotherPath', doAnotherThing)
]))
```

**Comment by a-s-o** [comment]
In your example, if an error occurs in async1 or async3 or setResults, it should go straight to setError and the chain should stop, which to me is the most predicable path.

**Comment by christianalfoni** [comment]
Hi again and thanks for the feedback!

Yes, the current implementation now basically does this:

1. If array of async actions, run them all and accumulate the results
2. Merge in arguments from results
3. Create array of paths that the actions resolved to
4. Merge paths into current execution

About async actions, I do not think of the paths as "success" or "error" in terms of "resolve" and "reject". They are just paths. Since you can define custom paths, it could have looked like this:

```js
controller.signal('appMounted',
  [async1, async2, async3],
  {
    foo: [setResults],
    bar: [setErrors]
  }
);
```

But seeing your example above with defining outputs on each async action, the idea of "collecting the paths" at the end is not a good idea.

```js
controller.signal('appMounted',
  [
    async1, {
      success: [],
      error: []
    },
    async2, {
      success: [],
      error: []
    }, 
    async3, {
      success: [],
      error: []
    }
  ]
);
```
It does look quite beautiful actually :-) And you can create smaller actions to handle each specific async action. What do you think?

So yeah, the way this would work is that each async action can choose any paths... it being the default ones or custom ones. If an async action has an uncaught exception it will throw to the console. If you want to handle exceptions you should use a try/catch and use the error path, or a custom path for it... in my opinion :-)

What do you think?

**Comment by a-s-o** [comment]
Makes sense. But I have a question so I understand bettwe. In your example you have, 
```javascript
[async1, { success, error }, async2, { success, error }]
```
should the action and the object dealing with its paths be wrapped in an array. For example,
```
[ [async1, { success, error }] , [ async2, { success, error } ] ]
  ^--- wrappers
```
because if not then in the following case is unclear to me:
```
[ async1, async2, { success, error } ]
                                    ^--- are these async2's paths or async1's
```
Sorry, a bit off topic but trying to understand what goes on implicitly in the code?

**Comment by christianalfoni** [comment]
That is a very good question!

To answer it with current mindset :-) :

```js
[ async1, async2, { success, error } ]
```
Now we say that "async1" does not resolve to any path. This means that async1 looks like this:

```js
const async1 = function (args, state, next) {
  next();
};
```
We say that "async2" does resolve to two possible paths. 

So if "async1" did `next.success()` or `next.error()` it would throw an error. If "async2" ran `next()` without a configured "defaultPath" it would throw an error. Does that make sense? :-)

**Comment by christianalfoni** [comment]
So it is the same as doing this:

```js
signal('name',
  mySyncAction,
  {
    success: [a],
    error: [b]
  },
  myOtherSyncAction
);
```
The only difference is that actions in an array runs async

**Comment by a-s-o** [comment]
Okay, makes sense. So the object only deals with the action that is directly before it. Thanks for clearing it up.

Back to topic, you said:
>I do not think of the paths as "success" or "error" in terms of "resolve" and "reject"

We should think of `error` as being the same as a `catch()` in promise land. `error` would be a special path. We can even rename it to `catch` to reduce confusion. The idea is that all other custom paths can be thought of as successful completion of the action by the signal except for error. Here's an example:
```javascript

const userLogin = cerebral.action({
   inputs: {
      username: Str,
      password: Str
   },
   outputs: {
      ok: {},
      accessDenied: { badLogin: Str },
      notFound: { badLogin: Str }
   }
   run: function userLogin (args, state, output) {
      const params = _.pick(args, 'username', 'password');
      http.post('/login', params. function callback (err, response) {
         if (err) {
            switch (err.statusCode) {

            // Both of these are "errors" of from the login perspective but 
            // as far as the signal is concerned the action was successfully
            // completed, so this result will get merged in args and the
            // chain will continue 
            case 401: return output.accessDenied({ badLogin: 'Access denied' });
            case 404: return output.notFound({ badLogin: 'User not found' });
            
            default:
               // Goes to the always present `error` path
               // or if not handled in the error path, it 
               // should stop the chain
               return output.error(err);
            }   
         }

         if (response.some_other_condition) {
            // Again goes to the `error` path, this is considered
            // a failure of the action, and if it is not handled it
            // stops the chain
            return output.error('Invalid response from server');
         }

         // Success
         output.ok();
      });
   }
})
```
So my logic is that the only time an action is considered to be `failed` is if it throws synchronously, or it calls `output.error`. So in effect `error` is the only special output path and it stops the signal chain. All other paths (custom, success, etc.), are considered a success. Hope that makes sense.

**Comment by a-s-o** [comment]
Here is an example of the above action being used
```javascript
const STOP = {};
// For now a simple way to stop the chain intentionally
function throwStop () {
   throw STOP;
}

cerebral.signal('loginFormSubmitted', ...[
   [userLogin, {
      accessDenied: [incrementCounter, displayMessage, throwStop],
      notFound: [incrementCounter, displayMessage, throwStop],
      lockedOut: [displayMessage, throwStop]
   }],

   // Successful login
   displayApp,
   [loadUserDocuments, loadAvatar],

   { 
      error: function (ex) {
         if (ex === STOP) return;

         // Do some other error recovery
      }
   }
]);
```
By the way, I think we are onto something. I stopped writing new actions while I wait for this new feature but as you said small actions with specific inputs and outputs will be very useful and portable. The only thing I can see comparable to this is using Bluebird's collection helpers, or async.series/waterfall, etc on node side but this may become very useful in the browser. Great work so far

**Comment by christianalfoni** [comment]
Hi @a-s-o and thanks for sticking with ideas you feel are important! :-)

I see your point with being able to stop the chain, though I think these cases can be solved using the current login. Like your example:

```js
cerebral.signal('loginFormSubmitted', ...[
   [userLogin, {
      success: [
        displayApp,
        [loadUserDocuments, loadAvatar]
      ], 
      accessDenied: [incrementCounter, displayMessage],
      notFound: [incrementCounter, displayMessage],
      lockedOut: [displayMessage]
   }],
```
I really want you to be able to start at the beginning of the signal and just read it through. Having this error output, which can be defined anywhere... and should propagate up the actions chain?... is a very complex implementation. But more importantly, it breaks the readability of the chain. You can not see from the signal definition that an action actually calls the error output.

So you could increase readability of where it throws, with for example:
```js
cerebral.signal('loginFormSubmitted', ...[
   [userLogin, {
      accessDenied: [incrementCounter, displayMessage,  new Error('error1')],
      notFound: [incrementCounter, displayMessage,  new Error('error2')],
      lockedOut: [displayMessage, new Error('error3')]
   }],
  displayApp,
  [loadUserDocuments, loadAvatar],
  {
    catch: [handleError]
  }
```
but you still have to find the catch, and there might be more of them. Comparing the first example above with this, the first one reads a lot better without any complex catch error propagation.

**Comment by a-s-o** [comment]
Hi Christian, you're right, much more readable. So in the first example, how do I get a hold of the errors. Let's say one of the following errors occurs:
```javascript
// Sync validation error
// User forgot to supply the number of submission so far (one of the incrementCounter 
// functions failed but we are not sure which one)
[incrementCounter:input] /timesSubmitted is undefined, should be a Num 

// Async error
// loadUserDocuments runs into a not implemented endpoint
Server returned a response of 412:NotImplemented
```
What would be the best place to catch these errors?

Really, the reason I am so concerned with error handling and doing it in the best possible way is that, it is so damn hard and so unpredictable. So, I like my tools to help me out and point out issues rather than a user finding them in a running app. Thanks for working through this with me

**Comment by christianalfoni** [comment]
Thanks to you too, really great input here :-)

There are two perspectives to this I think. One is that inputs/outputs helps you to define the relationship between actions. Cerebral will throw compile time errors when you are not using actions correctly and runtime errors when you run tests etc. It is a development tool.

The other perspective is that the inputs/outputs should be used for runtime errors in production, to recover from unexpected errors.

So currently I have been looking at this as a development tool, not an error handling tool in production. I think it is better to have a great development tool that gives good errors that can be reported from production and easily fixed, than having all sorts of fallbacks all over your code. Making it harder to read and reason about. An unexpected error is a bug, and it should be fixed :-)

So the first example indicates that there is a bug in the VIEW layer, as it was possible to trigger a signal with the wrong arguments. It is a bug that should be fixed.

The second example also indicates a bug that needs to be fixed. You can not really do anything for the user or the application to recover from it... as what you need does not exist.

That said! General errors with the flow of the application could of course be catched to act upon, like passing it to an error service like mixpanel, giving a message to the user that something bad happened and it has been reported.

```js
cerebral.on('error', function (error) {
  cerebral.signals.errorOccurred({error: error});
  mixpanel.send(error);
});
```

So yeah, quick summary. Better to report unexpected errors effectively and with good details, than having all sorts of fallback code spread around to catch bugs :-)

**Comment by a-s-o** [comment]
Right, I agree with that. I like this even better.

**Comment by christianalfoni** [comment]
Awesome! I am implementing as we speak, thanks so much for all the feedback. Will let you know when I get some commits up and we can try and document :-)

**Comment by christianalfoni** [comment]
Okay, I think the implementation is done. I have documented a bunch, but want to create a video introducing it all before releasing it. So I will do this tomorrow :-)

**Comment by a-s-o** [comment]
Look forward to it. Thanks for implementing this so fast

**Comment by christianalfoni** [comment]
Okay, got the new version out. I chose not to use tcomb as it had a few extra stuff we did not need and the validator was not intended to validate just one of many props, as the args object is.

But that led me to implementing a very simple "none dependency" type checker as a default and a function hook to use any types library you want.

**Comment by a-s-o** [comment]
Hi Christian. I saw the video and the new addition looks good. Thanks a lot for your hard work and incorporating the action ideas. I will get started and let you know how it goes.

**Comment by christianalfoni** [comment]
@a-s-o Cool! I was not extremely happy with the introduction clip, but yeah, not spending too much time on it as the API is still in flux. Make some better ones when things are more "locked down" :-)

Thanks for checking it out and I would happily receive documentation updates and maybe some error updates. Not sure about how those errors are shown, could be better. And of course make sure everything works :-)

Again, thanks for the inputs @a-s-o!
